### PR TITLE
[Polyfill] Allow to offload to WebNN backend

### LINKED ISSUE #594: [Polyfill] Allow to offload to WebNN backend
@pinzhenx , I captured the idea based on our brainstorming. Please take a look. Thanks!

The idea includes:
1. **Configuration:** mock a list of supported ops of WebNN backend (later WebNN can report supported ops)
2. **Compilation:** WASM polyfill partitions the model into sub-models based on the supported ops list
3. **Execution:** WASM executes the unsupported ops by WASM ops and offload sub-models to WebNN for acceleration

### PR DESCRIPTION
**Pull request by pinzhenx**
Fix #594

### CODE REVIEW

**Comment by pinzhenx** [comment]
@huningxin I think it would be better to create another branch for this feature. Because there are some temporary solutions involved in the polyfill, examples and benchmark. 
- Benchmark and examples used global variables to pass the `hybridPreferCode`, `isEagerMode` and `supportedOpList` into the polyfill. They will be removed once we have the introspection capability.
- Modified the `setPreference` interface to pass these parameters
- Added a new `OperationCode` called `NATIVE_OP`. This is also a temporary solution.
- Examples are only able to run in the Hybrid mode now, not in the pure WebNN or pure Polyfill mode

**Comment by pinzhenx** [comment]
@huningxin I think I should probably open another PR without demo code and profiling code in examples and benchmark, and use the `supportedOpsList` according to [docs/supported_ops.md](https://github.com/intel/webml-polyfill/blob/master/docs/supported_ops.md) for now

**Comment by huningxin** [comment]
> * They will be removed once we have the introspection capability

What is the introspection capability?

> * Added a new `OperationCode` called `NATIVE_OP`. This is also a temporary solution.

It seems OK, we may just rename it to `WEBNN_SUBGRAPH` for example.

> * Hybrid mode

Does it require a new build of polyfill?



> I think I should probably open another PR without demo code (supported ops selection and eager mode) and profiling code in examples and benchmark, and use the `supportedOpsList` according to [docs/supported_ops.md](https://github.com/intel/webml-polyfill/blob/master/docs/supported_ops.md) for now

The benchmark / demo with "supported ops select" and "eager mode" feature are useful. We should include them into the repo.

For `supportedOpsLis`, before WebNN supports reporting supported ops, we may need to set default values according to our implementation and expose a polyfill API to allow benchmark / demo to select.

**Comment by pinzhenx** [comment]
@huningxin @BruceDai  I managed to keep all the original logic of examples and benchmark. On top of that, I added the following features:

- API change
I found the `options` in the constructor of `Model` a good place to put all the extra parameters, such as `eager` and `supportedOps`, and it will only affect the polyfill.
```js
export default class Model {
  constructor(options = {}) {
    this._completed = false;
    this._operands = [];
    this._operations = [];
    this._inputs = null;
    this._outputs = null;
    this._backend = options.backend;
    this._eager = options.eager || false;
    this._supportedOps = options.supportedOps || new Set();
  }
}
```

- Hybrid execution
Allow running in either 'hybrid' mode or 'solo' mode. This is made possible by the `preference` and `supportedOps` in the polyfill.
If the `supportedOps` is not an empty set, the `preference` will be no longer a don't-care. So if the user still wants to use only the polyfill backend, just set the `supportedOps` to empty.

- Polyfill profiling
The partitioning will not display by default so that it won't pollute the log of examples. One has to explicitly call `dumpProfilingResults` in the prepared model, then it will output the subgraphs along with timing results, it is enabled in the benchmark.

**Comment by pinzhenx** [comment]
There is a small issue detected in the regression test. I will fix it soon

File: examples/image_classification/index.html [hunk 1]
```diff
@@ -144,6 +144,27 @@
     </div>
   </form>
 
+  <div class="container supported-ops-select">
+    <div class="row small">
+      <button class="col-4" onclick="defaultSupportedOps=getDefaultSupportedOps(currentBackend, currentPrefer);document.querySelectorAll('input[name=supportedOp]').forEach(x => x.checked=defaultSupportedOps.has(parseInt(x.value)))">Select Supported</button>
```
Comment (ibelem) [inline examples/image_classification/index.html]: How about to split such kind of "onclick" event codes into ui.js ?

**Comment by pinzhenx** [comment]
@ibelem Please take a look again. Thanks

**Comment by pinzhenx** [comment]
No regression found:

**WASM**:
passes: 508 failures: 262

**WebGL**:
passes: 499 failures: 271

**Comment by pinzhenx** [comment]
@ibelem  @ibelem  @BruceDai Please take a look, thanks.
- Add docs
- Remove ops selection in examples
- Add an interface in `utils` for getting required ops
   `utils.getRequiredOps` will return a promise that will be resolved once the `utils.init` is finished. The promise will be fulfilled with a set of supported ops code of type `Set<numer>`. 
   There's no UI using this interface yet.

**Comment by huningxin** [comment]
> Add docs
> * Graph partition
> * Subgraph construction
> * Enums.js - `WEBNN_SUBGRAPH`
> * Model.js - Model options
> * base.js

Looks good to me. Well done!

> * Add an interface in `utils` for getting required ops
>   `utils.getRequiredOps` returns a promise that will be resolved once the `utils.init` is finished. The promise will be fulfilled with a `Set<number>`, containing all supported ops code
> * Add an interface in `utils` for getting summary of subgraphs
>   `utils.getSubgraphsSummary` returns a list of subgraphs if the model is ready. Otherwise, it returns an empty list.

It looks like there are some duplicate code in each example. Can we consolidate them by sub-classing? @ibelem , what do you think?

**Comment by ibelem** [comment]
@huningxin Agree and i prefer to have seperate PR for the "duplicate code" issue.

**Comment by ibelem** [comment]
Merge this PR, more UI improvements and "duplicate code" issue will come.

**Pull request merged**